### PR TITLE
refactor:  update quota consumption method

### PR DESCRIPTION
**Pull request by flywukong**
### Description

1. change the algorithm of quota consumption
The freeQuota will no longer be reset and updated monthly. Instead, it will be fixed based on the value on the chain, and once it is fully consumed, it will not be available anymore.

3. support commands to update free quota
SP can use the command to update his free quota on greenfield

### Rationale

improve the way of quota consume

### Example

NA

### Changes

Notable changes: 
* update quota consumption method

### CODE REVIEW

File: base/gnfd/gnfd_service.go [hunk 1]
```diff
@@ -200,6 +200,38 @@ func (g *Gnfd) QuerySP(ctx context.Context, operatorAddress string) (*sptypes.St
...
+}
+
+// QuerySPPrice returns the sp price info
+func (g *Gnfd) QuerySPPrice(ctx context.Context, operatorAddress string) (sptypes.SpStoragePrice, error) {
+	startTime := time.Now()
+	defer metrics.GnfdChainTime.WithLabelValues("query_spPrice").Observe(time.Since(startTime).Seconds())
+	client := g.getCurrentClient().GnfdClient()
+	resp, err := client.QueryGetSpStoragePriceByTime(ctx, &sptypes.QueryGetSpStoragePriceByTimeRequest{
+		SpAddr:    operatorAddress,
+		Timestamp: 0,
```
Comment (sysvm) [inline base/gnfd/gnfd_service.go:226]: ditto

File: cmd/command/quota.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		return err
+	}
+
+	if oldQuota == freeQuota {
+		return fmt.Errorf("quota value: %d is same as chain meta, no need to update \n", freeQuota)
+	}
+
+	priceInfo, err := chain.QuerySPPrice(localCtx, operatorAddr)
+	if err != nil {
+		fmt.Printf("fail to get the current quota info %s\n", err.Error())
```
Comment (sysvm) [inline cmd/command/quota.go:59]: failed to get

File: modular/downloader/download_task.go [hunk 3]
```diff
@@ -207,34 +244,57 @@ func (d *DownloadModular) PreDownloadPiece(ctx context.Context, downloadPieceTas
...
-					ReadQuotaSize: downloadPieceTask.GetBucketInfo().GetChargedReadQuota() + d.bucketFreeQuota,
-				},
-			); err != nil {
-				metrics.PerfGetObjectTimeHistogram.WithLabelValues("get_object_check_quota_time").Observe(time.Since(checkQuotaTime).Seconds())
-				log.CtxErrorw(ctx, "failed to check bucket quota", "error", err)
-				if errors.Is(err, sqldb.ErrCheckQuotaEnough) {
-					return ErrExceedBucketQuota
-				}
-				// ignore the access db error, it is the system's inner error, will be let the request go.
+		log.CtxDebugw(ctx, "downloading from primary SP, checking quota")
```
Comment (sysvm) [inline modular/downloader/download_task.go:247]: use log.CtxDebug

File: core/spdb/entity.go [hunk 4]
```diff
@@ -57,17 +57,24 @@ type ReadRecord struct {
...
-	ReadConsumedSize uint64
-	ReadQuotaSize    uint64
-	ModifyTime       int64
+	BucketID              uint64
+	BucketName            string
+	ReadConsumedSize      uint64
+	FreeQuotaSize         uint64 // the total free quota size of SP price meta
+	FreeQuotaConsumedSize uint64 // the consumed free quota size
+	ChargedQuotaSize      uint64 // the total charged quota of bucket
+	ModifyTime            int64
```
Comment (sysvm) [inline core/spdb/entity.go:77]: use `ModifiedTime`

File: base/gnfd/gnfd_service.go [hunk 5]
```diff
@@ -200,6 +200,38 @@ func (g *Gnfd) QuerySP(ctx context.Context, operatorAddress string) (*sptypes.St
...
 }
 
+// QuerySPFreeQuota returns the sp free quota
+func (g *Gnfd) QuerySPFreeQuota(ctx context.Context, operatorAddress string) (uint64, error) {
+	startTime := time.Now()
+	defer metrics.GnfdChainTime.WithLabelValues("query_spQuota").Observe(time.Since(startTime).Seconds())
+	client := g.getCurrentClient().GnfdClient()
+	resp, err := client.QueryGetSpStoragePriceByTime(ctx, &sptypes.QueryGetSpStoragePriceByTimeRequest{
+		SpAddr:    operatorAddress,
+		Timestamp: 0,
```
[the remark this answers is not in the document]
Comment (flywukong) [reply base/gnfd/gnfd_service.go]: removed

File: base/gnfd/gnfd_service.go (hunk 1 above)
Comment (flywukong) [reply base/gnfd/gnfd_service.go]: fixed

File: cmd/command/quota.go (hunk 2 above)
Comment (flywukong) [reply cmd/command/quota.go]: fixed

File: modular/downloader/download_task.go (hunk 3 above)
Comment (flywukong) [reply modular/downloader/download_task.go]: remove the log

File: base/gnfd/gnfd_service.go [hunk 6]
```diff
@@ -200,6 +200,36 @@ func (g *Gnfd) QuerySP(ctx context.Context, operatorAddress string) (*sptypes.St
 	return resp.GetStorageProvider(), nil
 }
 
+// QuerySPFreeQuota returns the sp free quota
+func (g *Gnfd) QuerySPFreeQuota(ctx context.Context, operatorAddress string) (uint64, error) {
+	startTime := time.Now()
+	defer metrics.GnfdChainTime.WithLabelValues("query_spQuota").Observe(time.Since(startTime).Seconds())
```
Comment (will-2012) [inline base/gnfd/gnfd_service.go:206]: query_sp_quota.

File: base/gnfd/gnfd_service.go [hunk 7]
```diff
@@ -200,6 +200,36 @@ func (g *Gnfd) QuerySP(ctx context.Context, operatorAddress string) (*sptypes.St
...
+		log.Errorw("failed to query storage provider", "error", err)
+		return 0, err
+	}
+	return resp.GetSpStoragePrice().FreeReadQuota, nil
+}
+
+// QuerySPPrice returns the sp price info
+func (g *Gnfd) QuerySPPrice(ctx context.Context, operatorAddress string) (sptypes.SpStoragePrice, error) {
+	startTime := time.Now()
+	defer metrics.GnfdChainTime.WithLabelValues("query_spPrice").Observe(time.Since(startTime).Seconds())
```
Comment (will-2012) [inline base/gnfd/gnfd_service.go:221]: query_sp_pirce

File: test/e2e/spworkflow/e2e_test.sh [hunk 8]
```diff
@@ -178,6 +189,9 @@ function test_sp_exit() {
     ./gnfd-cmd -c ./config.toml --home ./ sp ls
     ./gnfd-cmd -c ./config.toml --home ./ bucket head gnfd://spexit
     ./gnfd-cmd -c ./config.toml --home ./ object head gnfd://spexit/example.json
+
+     ./gnfd-cmd -c ./config.toml --home ./ bucket get-quota  gnfd://spexit
```
Comment (will-2012) [inline test/e2e/spworkflow/e2e_test.sh:193]: Define test quota function.

File: modular/downloader/download_task.go [hunk 9]
```diff
@@ -43,6 +50,30 @@ func (d *DownloadModular) PreDownloadObject(ctx context.Context, downloadObjectT
...
+		if err != nil {
+			return ErrConsensus
+		}
+		// only need to set the free quota when init the traffic table
+		err = d.baseApp.GfSpDB().InitBucketTraffic(bucketID, bucketName, &spdb.BucketQuota{
+			ChargedQuotaSize: downloadObjectTask.GetBucketInfo().GetChargedReadQuota(),
+			FreeQuotaSize:    freeQuotaSize,
+		})
+		if err != nil {
+			log.CtxErrorw(ctx, "init bucket traffic fail", "error", err)
```
Comment (will-2012) [inline modular/downloader/download_task.go:72]: failed to init.

File: modular/downloader/download_task.go [hunk 10]
```diff
@@ -207,34 +244,56 @@ func (d *DownloadModular) PreDownloadPiece(ctx context.Context, downloadPieceTas
...
+		if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
+			return err
+		}
+		// init the bucket traffic table when checking quota for the first time
+		if bucketTraffic == nil {
+			freeQuotaSize, err = d.baseApp.Consensus().QuerySPFreeQuota(ctx, d.baseApp.OperatorAddress())
+			if err != nil {
+				return ErrConsensus
 			}
+			log.CtxDebugw(ctx, "finish init bucket traffic table", "charged quota:", downloadPieceTask.GetBucketInfo().GetChargedReadQuota(),
```
Comment (will-2012) [inline modular/downloader/download_task.go:261]: Refine "charged quota:" to "charged_quota"

File: modular/downloader/download_task.go [hunk 11]
```diff
@@ -207,34 +244,56 @@ func (d *DownloadModular) PreDownloadPiece(ctx context.Context, downloadPieceTas
...
+				"freeQuota", freeQuotaSize)
+
+			// only need to set the free quota when init the traffic table
+			err = d.baseApp.GfSpDB().InitBucketTraffic(bucketID, bucketName, &spdb.BucketQuota{
+				ChargedQuotaSize: downloadPieceTask.GetBucketInfo().GetChargedReadQuota(),
+				FreeQuotaSize:    freeQuotaSize,
+			})
+			if err != nil {
+				log.CtxErrorw(ctx, "init bucket traffic fail", "error", err)
```
Comment (will-2012) [inline modular/downloader/download_task.go:270]: failed to init

File: modular/gater/bucket_handler.go [hunk 12]
```diff
@@ -62,11 +66,27 @@ func (g *GateModular) getBucketReadQuotaHandler(w http.ResponseWriter, r *http.R
 		err = ErrConsensus
 		return
 	}
-	charge, free, consume, err = g.baseApp.GfSpClient().GetBucketReadQuota(
-		reqCtx.Context(), bucketInfo, reqCtx.vars["year_month"])
-	if err != nil {
-		log.CtxErrorw(reqCtx.Context(), "failed to get bucket read quota", "error", err)
-		return
+
+	log.CtxDebugw(reqCtx.Context(), "get bucket info when query quota:", "quota", bucketInfo.ChargedReadQuota)
```
Comment (will-2012) [inline modular/gater/bucket_handler.go:70]: refine `quota:` to `quota`.

File: modular/gater/bucket_handler.go [hunk 13]
```diff
@@ -62,11 +66,27 @@ func (g *GateModular) getBucketReadQuotaHandler(w http.ResponseWriter, r *http.R
...
+	bucketTraffic, err = g.baseApp.GfSpDB().GetBucketTraffic(
+		bucketInfo.Id.Uint64())
+	// if the traffic table has not been created and initialized yet, return the chain info
+	if errors.Is(err, gorm.ErrRecordNotFound) || bucketTraffic == nil {
+		free, err = g.baseApp.Consensus().QuerySPFreeQuota(reqCtx.Context(), g.baseApp.OperatorAddress())
+		if err != nil {
+			err = ErrConsensus
+		}
+		charge = bucketInfo.GetChargedReadQuota()
+		log.CtxDebugw(reqCtx.Context(), "fail to get traffic info in db, return the chain meta", "free quota", free, "charged quota", charge)
```
Comment (will-2012) [inline modular/gater/bucket_handler.go:80]: `"free quota", free, "charged quota", charg` to `"free_quota", free, "charged_quota", charge`

**Review by will-2012 — COMMENTED** [review]
Please check failed action

File: modular/gater/bucket_handler.go (hunk 13 above)
Comment (flywukong) [reply modular/gater/bucket_handler.go]: updated

File: modular/gater/bucket_handler.go (hunk 12 above)
Comment (flywukong) [reply modular/gater/bucket_handler.go]: fixed

File: modular/downloader/download_task.go (hunk 10 above)
Comment (flywukong) [reply modular/downloader/download_task.go]: fixed

File: test/e2e/spworkflow/e2e_test.sh (hunk 8 above)
Comment (flywukong) [reply test/e2e/spworkflow/e2e_test.sh]: removed

**Review by RenRick — APPROVED** [review]
LGTM

**Review by sysvm — APPROVED** [review]
LGTM

**Pull request merged**